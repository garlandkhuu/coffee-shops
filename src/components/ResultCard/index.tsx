import { forwardRef, ReactNode } from 'react';

import {
  Container,
  Thumbnail,
  InfoContainer,
  Info,
  Name,
  Pricing,
} from './ResultCard.styles';
import { CoffeeShopListResult } from '../../interfaces';
import assemblePhotoUrl from '../../utils/assemblePhotoUrl';
import parsePrice from '../../utils/parsePrice';
import defaultImage from '../../assets/no-image.png';
import useResponsive from '../../hooks/useResponsive';

interface ResultCardProps {
  children?: ReactNode;
  shop: CoffeeShopListResult;
  viewDetails: (placeId: string) => void;
}

const ResultCard = forwardRef<HTMLDivElement, ResultCardProps>(
  ({ shop, viewDetails }, ref) => {
    const { name, price, rating, fsq_id: placeId } = shop;

    const { location } = shop;
    const { address } = location;

    const { photos } = shop;
    const thumbnail = photos[0];
    const prefix = thumbnail ? thumbnail.prefix : '';
    const suffix = thumbnail ? thumbnail.suffix : '';

    const { isTablet, isDesktop } = useResponsive();

    return (
      <Container onClick={() => viewDetails(placeId)} ref={ref}>
        <Thumbnail
          src={
            prefix && suffix
              ? assemblePhotoUrl(prefix, suffix, '250x250')
              : defaultImage
          }
        />
        <InfoContainer>
          <Name>{name}</Name>
          <Info>
            <p>
              {address && `${address}`}
              {isDesktop && <Pricing> · {parsePrice(price)}</Pricing>}
            </p>
            <p>
              Rating: {rating ? `${rating}/10` : 'Unavailable'}
              {isTablet && <Pricing> · {parsePrice(price)}</Pricing>}
            </p>
          </Info>
        </InfoContainer>
      </Container>
    );
  }
);

export default ResultCard;
