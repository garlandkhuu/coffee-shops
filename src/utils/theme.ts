export interface Theme {
  colors: {
    beigeLight: string;
    brown: string;
    brown2: string;
    black: string;
    blackOpaque: string;
    white: string;
    grey: string;
    grey2: string;
    sage: string;
    red: string;
  };
  fonts: {
    regular: string;
    bold: string;
    decorated: string;
  };
  maxPageWidth: number;
}

const theme: Theme = {
  colors: {
    beigeLight: '#e2d5ca',
    brown: '#492e22',
    brown2: '#a6816a',
    black: '#000000',
    blackOpaque: 'rgba(0, 0, 0, 0.5)',
    white: '#FFFFFF',
    grey: '#dbd8d7',
    grey2: '#969696',
    sage: '#209971',
    red: '#d43d69',
  },
  fonts: {
    regular: 'Lato',
    bold: 'Lato',
    decorated: 'Besley',
  },
  maxPageWidth: 1500,
};

export default theme;
