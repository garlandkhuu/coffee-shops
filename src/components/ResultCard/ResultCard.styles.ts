import styled from 'styled-components';

import { MEDIA_QUERIES } from '../../utils/constants';

export const Container = styled.div`
  display: flex;
  flex-direction: row;
  margin: 0 15px 25px;
  border-radius: 12px;
  background: ${({ theme }) => theme.colors.white};
  cursor: pointer;
  width: calc(50% - 30px);
  max-height: 250px;
  box-shadow: ${({ theme }) => `3px 3px 10px 3px ${theme.colors.grey}`};

  ${MEDIA_QUERIES.medium} {
    max-height: 125px;
    width: 100%;
  }
`;

export const Thumbnail = styled.img`
  object-fit: cover;
  border-radius: 12px 0 0 12px;
  width: 250px;
  height: 250px;

  ${MEDIA_QUERIES.medium} {
    width: 125px;
    height: 125px;
  }
`;

export const InfoContainer = styled.div`
  display: flex;
  flex-direction: column;
  padding: 25px 20px;

  ${MEDIA_QUERIES.medium} {
    padding: 15px 20px;
  }
`;

export const Info = styled.div`
  display: flex;
  flex-direction: column;
  height: 100%;
  justify-content: space-between;

  p:last-child {
    margin-bottom: 0;
  }

  ${MEDIA_QUERIES.medium} {
    p:first-child {
      margin: 10px 0 0 0;
    }

    p:last-child {
      margin-top: 10px;
    }

    p: {
      font-size: 14px;
    }
  }
`;

export const Name = styled.h3`
  margin: 0 0 10px 0;

  ${MEDIA_QUERIES.medium} {
    margin: 0;
    font-size: 18px;
  }
`;

export const Pricing = styled.span`
  font-family: ${({ theme }) => theme.fonts.bold};
  font-weight: bold;
  letter-spacing: 2px;
  color: ${({ theme }) => theme.colors.sage};
`;
