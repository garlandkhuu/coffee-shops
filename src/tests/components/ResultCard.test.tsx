import 'jsdom-global/register';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import configureStore from 'redux-mock-store';
import { shallowToJson } from 'enzyme-to-json';
import 'jest-styled-components';

import ResultCard from '../../components/ResultCard';
import theme from '../../utils/theme';
import { mount, shallow } from '../setup/test-setup';
import { mockShopResult } from '../__mocks__/data';

const mockStore = configureStore();
const mockViewDetails = jest.fn();

describe('ResultCard', () => {
  let component: JSX.Element;

  beforeEach(() => {
    component = (
      <ThemeProvider theme={theme}>
        <Provider store={mockStore()}>
          <ResultCard shop={mockShopResult} viewDetails={mockViewDetails} />
        </Provider>
      </ThemeProvider>
    );
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should match snapshot', () => {
    const wrapper = shallow(component);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should render details properly', () => {
    const wrapper = mount(component);
    const name = wrapper.find('h3');
    expect(name.text()).toEqual('Stumptown Coffee Roasters');

    const price = wrapper.find('span').first();
    expect(price.text()).toEqual(' Â· $');
  });
});
