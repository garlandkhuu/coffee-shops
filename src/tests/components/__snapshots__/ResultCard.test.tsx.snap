// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultCard should match snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "colors": Object {
        "beigeLight": "#e2d5ca",
        "black": "#000000",
        "blackOpaque": "rgba(0, 0, 0, 0.5)",
        "brown": "#492e22",
        "brown2": "#a6816a",
        "grey": "#dbd8d7",
        "grey2": "#969696",
        "red": "#d43d69",
        "sage": "#209971",
        "white": "#FFFFFF",
      },
      "fonts": Object {
        "bold": "Lato",
        "decorated": "Besley",
        "regular": "Lato",
      },
      "maxPageWidth": 1500,
    }
  }
>
  <Provider
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <ForwardRef
      shop={
        Object {
          "categories": Array [
            Object {
              "icon": Object {
                "prefix": "https://ss3.4sqi.net/img/categories_v2/food/burger_",
                "suffix": ".png",
              },
              "id": 13031,
              "name": "Burger Joint",
            },
            Object {
              "icon": Object {
                "prefix": "https://ss3.4sqi.net/img/categories_v2/food/cafe_",
                "suffix": ".png",
              },
              "id": 13034,
              "name": "CafÃ©",
            },
            Object {
              "icon": Object {
                "prefix": "https://ss3.4sqi.net/img/categories_v2/food/coffeeshop_",
                "suffix": ".png",
              },
              "id": 13035,
              "name": "Coffee Shop",
            },
          ],
          "fsq_id": "49c2e861f964a5203f561fe3",
          "location": Object {
            "address": "1022 SW Harvey Milk St",
            "country": "US",
            "cross_street": "at SW 11th Ave.",
            "dma": "Portland, Or",
            "formatted_address": "1022 SW Harvey Milk St (at SW 11th Ave.), Portland, OR 97205",
            "locality": "Portland",
            "neighborhood": Array [
              "Chinatown",
            ],
            "postcode": "97205",
            "region": "OR",
          },
          "name": "Stumptown Coffee Roasters",
          "photos": Array [
            Object {
              "created_at": "2012-05-08T01:05:51.000Z",
              "height": 540,
              "id": "4fa8716fe4b0f380f2a34417",
              "prefix": "https://fastly.4sqi.net/img/general/",
              "suffix": "/nr5BD9e-ugF4OlKEr3kekomATxSlhAk9uUMfbgFms48.jpg",
              "width": 720,
            },
            Object {
              "classifications": Array [
                "food",
              ],
              "created_at": "2012-03-14T23:13:40.000Z",
              "height": 540,
              "id": "4f612624e4b0deb11b0d5119",
              "prefix": "https://fastly.4sqi.net/img/general/",
              "suffix": "/QzWHmQZu_TIGm8yTBiMACU1dU9QQARt80YFdsj5y37I.jpg",
              "width": 720,
            },
            Object {
              "created_at": "2012-12-30T18:56:57.000Z",
              "height": 612,
              "id": "50e08e79e4b0cf8b435cc233",
              "prefix": "https://fastly.4sqi.net/img/general/",
              "suffix": "/5081071_D9R6MynBglWCG_dwE5HoWnO0KIILCDxovdpS-MrvErw.jpg",
              "width": 612,
            },
            Object {
              "created_at": "2013-05-09T15:51:23.000Z",
              "height": 1920,
              "id": "518bc5fb498ec40717d07446",
              "prefix": "https://fastly.4sqi.net/img/general/",
              "suffix": "/3093239_zhAhqFc8vKh3AmbzTl8I1H_dz4Uo2U22FRxL2aFWf9A.jpg",
              "width": 1440,
            },
            Object {
              "created_at": "2012-05-03T18:30:11.000Z",
              "height": 537,
              "id": "4fa2ceb3e4b01ab96b096506",
              "prefix": "https://fastly.4sqi.net/img/general/",
              "suffix": "/z0asVCOU40wzBcoaumuVxRO6yEGS93Mg8elWn8mJRfs.jpg",
              "width": 720,
            },
          ],
          "price": 1,
          "rating": 9,
        }
      }
      viewDetails={[MockFunction]}
    />
  </Provider>
</ContextProvider>
`;
